test/e2e/app-dir/ppr-full/components/dynamic.jsx

```diff
@@ -1,37 +1,38 @@
-import React from 'react'
-import { headers } from 'next/headers'
+import React, { use } from 'react'
+import * as next from 'next/headers'
 
 export const Dynamic = ({ pathname, fallback }) => {
   if (fallback) {
-    return <div>Loading...</div>
+    return <div>Dynamic Loading...</div>
   }
 
+  const headers = next.headers()
   const messages = []
-  const names = ['x-test-input', 'user-agent']
-  const list = headers()
+  for (const name of ['x-test-input', 'user-agent']) {
+    messages.push({ name, value: headers.get(name) })
+  }
 
-  for (const name of names) {
-    messages.push({ name, value: list.get(name) })
+  const delay = headers.get('x-delay')
+  if (delay) {
+    use(new Promise((resolve) => setTimeout(resolve, parseInt(delay, 10))))
```

Comment:
Not worth blocking the PR, but: I don't love the use of timers to detect I/O because it's inherently a bit flaky (something unrelated could cause it to be too slow) and it makes the tests run slower.

What I'm planning to do for testing client navigations is spin up an HTTP server and push to an event log whenever the app sends a request, like this:

https://github.com/vercel/next.js/blob/6b3d145e7e7e88037f4e326cc63038814ab5fd76/test/e2e/app-dir/ppr-navigations/ppr-navigations.test.ts#L19-L80

I like the event log pattern for testing because it also detects things like if the same fetch happens more than once, which is useful for catching caching bugs.

Reply:
Specifically the delay is to test the "sending the static part of the payload first" behavior of PPR. From my understanding, it's the most black-box way of determining if the static part was cached + sent versus regenerated. I'm not too sure how the logs accomplish the same thing.

I could see using a server like this to instead to send a specific "random" value for it to use (the static value) and then a random value on future renders to help distinguish between what was rendered + returned statically and what was streamed dynamically without reaching into the implementation of it.